### ISSUE TITLE
Switch to Kotlin

### ISSUE THREAD

**Issue by RyanTheAllmighty** [ATLauncher/ATLauncher#695]
Switching to Kotlin would be nice, ideally we can do a staggered move, move over a couple things and then go on from there, but not sure if possible.

Main reason I'm more motivated now as moving more things over to GraphQL, the Apollo library we use we're currently a version behind because of lacklustre Java support, plus async/await is nice and Kotlin as a whole is just nicer.

Would be good to get some thought from @Doomsdayrs on this

**Comment by Doomsdayrs** [comment]
Good morning!

Ah darn, there goes my work on the VanillaPacksTab! 🥲

Regardless, Kotlin is possible, but there are a few things to note.

1. Migration will take time. Yes we can do a "rip the bandaid off" and convert the entire project to Kotlin at once. But there will be numerous logical issues, such as Kotlin being null safe, which means intensive look over the entire codebase. This is why if you notice in all of my recent PRs, I have always tried to include `@Nullable` & `@NotNull` flags. 
2. Kotlin is not a magical cure all. Yes it provides async functionalities & lowered header costs, but that does not help if the *architecture* of the launcher isn't well designed for it. My work on the view models are precisely one of the reasons for this (#606), As seen in the now defunct #659.

**Comment by Doomsdayrs** [comment]
If my VanillaPacksTab gets merged this week, we will have the first steps into kotlin. I am very close to fixing it up (Please do not change anything with VanillaPacksTab again).

`com.atlauncher.utils.Pair` can be the next thing to be axed, as Kotlin has pairs provided built into the language.

**Comment by Doomsdayrs** [comment]
One shenanigan that can be done separately from the code base is the build script, a migration to kts is possible.

I'll look into it when I have time

**Comment by RyanTheAllmighty** [comment]
The changes for this were reverted, it was attempted, but led to some issues with VSCode due to Kotlin imports in Java not being supported (https://www.turtlestoffel.com/the-state-of-kotlin-and-vscode/).

Ideally I'd love to have this working, but it just doesn't work as it is currently, and the advantages it gives over not being able to be productive in the repo

**Comment by ellet0** [comment]
I'm with this change

> Switching to Kotlin would be nice, ideally we can do a staggered move, move over a couple things and then go on from there, but not sure if possible.

You can add and start using Kotlin code right away without affecting the Java code in any way, in fact, Kotlin/JVM was designed to work with Java and you can use anything from Java in Kotlin and vice versa but you might already know that, we don't have to migrate everything right away, only the things we work on and the the new features and code and it will be migrated completely to Kotlin without affecting the development time

Kotlin will also allow you to use modern language features like a sealed class in Java 8 without requiring Java 17, the sealed class was introduced in Kotlin before Java and to me, it has better syntax, has many built-in functions so we don't create new functions all over the place that would make maintaining things harder and all developers have to be familiar with our custom code, also it designed with performance in mind, increase productivity, perfect interoperability with Java, better readability and maintainability, less runtime exceptions, null safety which is important to catch earlier bugs or handle them in effective way, strong IntelliJ IDEA support, much easier to get started allow more developers to contribute and many other features more than coroutines (async/await) like extension functions and it's just keep getting more and more features and more releases and support and the community is growing really fast, there is already support in VS Code for Kotlin (see [this](https://kotlinlang.org/docs/kotlin-ide.html/))

All we have to do is add Kotlin JVM plugin in Gradle and get started

**Comment by Doomsdayrs** [comment]
> I'm with this change

While the support is appreciated, even I recognize that whats important is ensuring that the maintainers development flow is not interrupted.

In this case, the more important task is architectural improvements of the application, instead of language improvements (#606)

**Comment by ellet0** [comment]
> > I'm with this change
> 
> While the support is appreciated, even I recognize that whats important is ensuring that the maintainers development flow is not interrupted.
> 
> In this case, the more important task is architectural improvements of the application, instead of language improvements (#606)

Both are important to create scalable and maintainable applications, as I mentioned, we don't have to migrate the full source code to Kotlin right away, Kotlin/JVM (even Java but in Kotlin it's more modern) has tons of libraries that force to separate the business logic from UI (separation of concern) 

It shouldn't interrupt anything, and I confirmed that I was able to add support for using Kotlin just by one line of code (enable Kotlin JVM Gradle plugin) and tested it out and everything works like expected, the only downside is 1MB larger bundle size which is worth what the library is providing, let me give me an example as proof

- If you take a look at this #893

In Java the code would be:

```java

// Joining a server on launch
        if (instance.launcher.initialJoinServerAddress != null && !instance.launcher.initialJoinServerAddress.isEmpty()) {
            String enteredServerAddress = instance.launcher.initialJoinServerAddress;
            boolean hasQuickMultiplayer = instance.arguments.game.stream().anyMatch(
                argumentRule -> argumentRule.value instanceof List &&
                    ((List<?>) argumentRule.value).contains("--quickPlayMultiplayer")
            );
            if (hasQuickMultiplayer) {
                // Minecraft 23w14a and newer versions
                arguments.add("--quickPlayMultiplayer");
                arguments.add(enteredServerAddress);
            } else {
                // Minecraft 23w13a and older versions
                String[] parts = enteredServerAddress.contains(":") ? enteredServerAddress.split(":")
                    : new String[]{enteredServerAddress};
                String address = parts[0];
                String port = parts.length > 1 ? parts[1] : String.valueOf(Constants.MINECRAFT_DEFAULT_SERVER_PORT);
                arguments.addAll(Arrays.asList("--server", address));
                arguments.addAll(Arrays.asList("--port", port));
            }
        }

```

In Kotlin it could be something like:

```kotlin

// Code migration from Java to Kotlin automated because this is just an example

instance.launcher.initialJoinServerAddress?.let { enteredServerAddress ->
    val hasQuickMultiplayer = instance.arguments.game.any {
        it.value is List<*> && (it.value as List<*>).contains("--quickPlayMultiplayer")
    }
    if (hasQuickMultiplayer) {
        // Minecraft 23w14a and newer versions
        arguments += "--quickPlayMultiplayer"
        arguments += enteredServerAddress
    } else {
        // Minecraft 23w13a and older versions
        val (address, port) = enteredServerAddress.split(":") + listOf(Constants.MINECRAFT_DEFAULT_SERVER_PORT.toString())
        arguments += "--server" to address
        arguments += "--port" to port
    }
}


```

You can't use many basic language features in Java unless you want to require the user to install a more recent version of Java (17 or 21) which is one of the reasons why newer versions of Minecraft require a more recent version of Java

But you probably already know all of this.

I suggest what I see best, not necessarily the right thing to do

**Comment by Doomsdayrs** [comment]
> But you probably already know all of this.

Yes, Such improvements been done in the past via #717. And more exist somewhere on my branch / local copies.

> Also using Kotlin would allow us to use...

@ellet0 While Compose would enable a far newer tech stack.

The problem is that the current stack is *utterly* outdated (SWING), as in its not even supported at all anymore by any one on this planet.

Business code and SWING UI code are completely integrated in many parts of the application.

I really suggest looking at what #606 has been trying to achieve, and what I have worked on for 2 years now.

If #606 is implemented, *anything* can be done, *any* new UI, *any* new programming language, etc, can be done far more easily.

> I suggest what I see best, not necessarily the right thing to do

You remind me of myself once upon a time, and I appreciate that you put a smile on my face today.

With projects like this, long term stability is more worth it then anything else (I've even considered writing a competitor application from scratch in Kotlin.).

Even my changes for future stability and modernization are likely quite annoying for @RyanTheAllmighty since the behavior changes etc.

**Comment by ellet0** [comment]
> You remind me of myself once upon a time, and I appreciate that you put a smile on my face today.

You too.

**Comment by Doomsdayrs** [comment]
> What architecture do you suggest to use? Is it MVVM?

Yes, MVC is more applicable for other situations. For a large project such as ATLauncher, MVVM is optimal.

**Comment by Doomsdayrs** [comment]
> Can you give me the directories where it has most of the UI that has business logic in it?

More discussion should be had in: https://github.com/ATLauncher/ATLauncher/issues/606

It already includes some information.

**Comment by RyanTheAllmighty** [comment]
Putting this back to high priority and "unblocking" this as it were. VSCode is just getting worse and worse over time when trying to deal with just a plain Java codebase, so I have switched to IntelliJ for most my work on the launcher.

The way I envision Kotlin sitting in with other things is:

- Current release needs to go out that's in master, this is not blocked anymore, but I'm waiting to do it on a weekend when I can pay more attention to any issues that pop up as that release waiting to go out has a lot of architectural stuff in it
- The mac app needs to be sorted out #502 with bundled JRE support
- Java 17 support #601 needs the last phase rolled out after the next release (this will bug people every single launch if not on Java 17)
- Then I think we can dive back into Kotlin migration, starting with undoing the reverts I did previous (where possible)
- Need to consider looking at Compose or JavaFX potentially, weigh up the benefits and see if it's worth it
- At some point there needs to be a cut off and switch to Java 17 for all
